core/renderers/measurables/rows.js

```diff
@@ -0,0 +1,389 @@
+    var elem = this.elements[i];
+    if (elem.isSpacer()) {
+      continue;
+    }
+    if (elem.isInput) {
+      return /** @type {Blockly.blockRendering.Input} */ (elem);
+    } else if (elem.isField()) {
+      return /** @type {Blockly.blockRendering.Input} */ (elem.parentInput);
+    }
+  }
+  return null;
+};
+
+/**
+ * Convenience method to get the first spacer element on this row.
+ * TODO: I want to say that this is guaranteed to exist, but I'm not sure how.
+ * I could check if it's null and throw, but that seems like overkill.
+ * @return {Blockly.blockRendering.InRowSpacer} The first spacer element on
+ *   this row.
+ * @package
+ */
+Blockly.blockRendering.Row.prototype.getFirstSpacer = function() {
+  return /** @type {Blockly.blockRendering.InRowSpacer} */ (this.elements[0]);
+};
+
+/**
+ * Convenience method to get the last spacer element on this row.
+ * TODO: I want to say that this is guaranteed to exist, but I'm not sure how.
+ * I could check if it's null and throw, but that seems like overkill.
+ * @return {Blockly.blockRendering.InRowSpacer} The last spacer element on
+ *   this row.
+ * @package
+ */
+Blockly.blockRendering.Row.prototype.getLastSpacer = function() {
+  /* eslint-disable-next-line max-len */
+  return /** @type {Blockly.blockRendering.InRowSpacer} */ (this.elements[this.elements.length - 1]);
+};
+
+/**
+ * An object containing information about what elements are in the top row of a
+ * block as well as sizing information for the top row.
+ * Elements in a top row can consist of corners, hats, spacers, and previous
+ * connections.
+ * @param {!Blockly.BlockSvg} block The block for which this represents the top
+ *     row.
+ * @package
+ * @constructor
+ * @extends {Blockly.blockRendering.Row}
+ */
+Blockly.blockRendering.TopRow = function(block) {
+  Blockly.blockRendering.TopRow.superClass_.constructor.call(this);
+
+  this.type = 'top row';
+
+  /**
+   * The starting point for drawing the row, in the y direction.
+   * This allows us to draw hats and simliar shapes that don't start at the
+   * origin. Must be non-negative (see #2820).
+   * @package
+   * @type {number}
+   */
+  this.startY = 0;
+
+  /**
+   * Whether the block has a previous connection.
+   * @package
+   * @type {boolean}
+   */
+  this.hasPreviousConnection = !!block.previousConnection;
+
+  /**
+   * The previous connection on the block, if any.
+   * TODO: Should this be the connection measurable instead? It would add some
+   * indirection but would mean we aren't mixing connections and connection
+   * measurables.
+   * @type {Blockly.RenderedConnection}
+   */
+  this.connection = block.previousConnection;
+
+  var precedesStatement = block.inputList.length &&
+      block.inputList[0].type == Blockly.NEXT_STATEMENT;
+
+  // This is the minimum height for the row. If one of its elements has a
+  // greater height it will be overwritten in the compute pass.
+  if (precedesStatement && !block.isCollapsed()) {
```

Comment:
TODO (fenichel): See if I can move the minimum height and maximum width stuff into measure, for both top and bottom rows.